<a href="https://colab.research.google.com/github/surya81/Alma-better-projects/blob/main/Telecom_Churn_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project name - Telecom Churn Analysis**

#### **Type** - EDA
#### **Contribution** - **Individual**

# **Problem Analysis**

 **Every telecom service provider company in course of development to adding new customers makes sures that they donot loose their existing customers .However every year a large no of people using their services changes from one provider to another in search of better services and price feasibilty,In oerder to bring back these existing customer an company eventually found themselves in expending a lot off costs**
**So,In order to maintain and develop better plans telecom compamnies does regular churn analysis of their customers for taking feedback in their lacking services and how to do better on these feedbackstext**

# **Objective**

**we will try to absorb valuable information from the customer data and we will use this information  to help in developing some new plans and method which will automitically leads us in reducing customer churn rate of the company**

#**Libraries**

In [2]:
import numpy as np
import pandas as pd 
import seaborn as sns

#**Dataset Loading**

In [3]:
def DatasetLoader(path): # A function to load the dataset 
    """
    Argument - path to the dataset 
    this function the the csv file using pandas function 
    return dataframe  """

    try :
        df = pd.read_csv(path) # read the data in csv format
        return df

    except: # if the path is not correct
        print("the specified file path is not correct")


In [4]:
telecom_df = DatasetLoader('/content/Telecom Churn.csv')#store the dataframe in telecom_df variable

# **Dataset Exploration**

In [5]:
telecom_df.head() #observe the first 5 rows of dataframe

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [6]:
print(f"the telecom  dataset has {telecom_df.shape[0]} rows and {telecom_df.shape[1]} columns ")

the telecom  dataset has 3333 rows and 20 columns 


#### **Dataset information**

In [7]:
telecom_df.info() # some basic information about the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

#### **Missing Values in Dataset**

In [8]:
telecom_df.isna().sum() # sum of missing values throughout each rows in dataset

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

#### **duplicated rows in dataset**

In [9]:
# Dataset Duplicate Value Count
print(len(telecom_df[telecom_df.duplicated()]))

0


**The dataset has total of 19 columns out of which 3 are object dtype , 8 are of float and int each and the last column which is also the target variable is bool type**

**Fortunately the data donot have any null values and any duplicated rows out of total 3333 rows and 19 columns**


# **Column description**



**State** - 51 Unique states of united states of america 

**Account length** - account length of each customer 

**Area Code** - 3 different area code for collectively all customer 

**International plan** - whethear a customer opted for internartional plan or not 

**Voicemail Plan** - wheather a customer has active voice mail plan

**number vmail message** - Number of voicemail message of each customer ranging from 0 to 50 

**Total day minutes** - duration of call made during morning 

**Total day calls** - total number of made during morning 

**Total day charge** - total morning charge to the customer

**Total eve minutes** - duration of call made during evening 

**Total eve calls** - total number of made during evening 

**Total evening charge** - total evening charge to the customer

**Total nigt minutes** - duration of call made during night

**Total night calls** - total number of made during night 

**Total night charge** - total morning charge to the night 

**churn** - wheather a customer opted for change in its telecom False for no  True for yes 



#### **total no of unique element in each column**

In [10]:
# function to count and print  unique no of elements in dataframe
def count_unique(dataframe): 
    """ 
    Argument - dataframe of the consideration 

    this function prints no of unique elements in each column 
    """
    try: 
        for columns in dataframe.columns.tolist():
            print(f"Total unque no of row in {columns} is {dataframe[columns].nunique()} ")

    except: #if dataframe does not exist 
        print("provided dataframe does not exist ")

In [11]:
# calling the function 
count_unique(telecom_df)

Total unque no of row in State is 51 
Total unque no of row in Account length is 212 
Total unque no of row in Area code is 3 
Total unque no of row in International plan is 2 
Total unque no of row in Voice mail plan is 2 
Total unque no of row in Number vmail messages is 46 
Total unque no of row in Total day minutes is 1667 
Total unque no of row in Total day calls is 119 
Total unque no of row in Total day charge is 1667 
Total unque no of row in Total eve minutes is 1611 
Total unque no of row in Total eve calls is 123 
Total unque no of row in Total eve charge is 1440 
Total unque no of row in Total night minutes is 1591 
Total unque no of row in Total night calls is 120 
Total unque no of row in Total night charge is 933 
Total unque no of row in Total intl minutes is 162 
Total unque no of row in Total intl calls is 21 
Total unque no of row in Total intl charge is 162 
Total unque no of row in Customer service calls is 10 
Total unque no of row in Churn is 2 


#### **description of columns**

In [12]:
telecom_df.describe() # description of numerical columns in the dataframe 

Account length    Area code  Number vmail messages  Total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       Total eve charge  Total night minutes  Total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       Total night charge  Total intl minutes  Total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       Total intl charge  Customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [13]:
#print the element name where no of unique elements in columns are less than or eaqual to three 
def display_unique_elements(dataframe):
    """ 
    Argument - dataframe we wish to know unique elements about 
    this function uinque elements where no unique elements are less than 3 
    return - none 
    """
    try :
        for columns in dataframe.columns.tolist():#converts column array to  a list 
            if dataframe[columns].nunique() <= 3 :
                print(f" the unique element in {columns} columns are  {dataframe[columns].unique()}")
    except :
        print('invalid dataframe')

In [14]:
#call function 
display_unique_elements(telecom_df)

 the unique element in Area code columns are  [415 408 510]
 the unique element in International plan columns are  ['No' 'Yes']
 the unique element in Voice mail plan columns are  ['Yes' 'No']
 the unique element in Churn columns are  [False  True]


**out of the four columns whose no of unique elements are less than or equal to 3 . 3 are in yes/no format we can change these with 0 and 1 . 1 for true and 0 for false**

#**Dataset manipulation**

####**Copy original dataset**

#### **Dataset column unique variable replace**

In [15]:
#function to replace variable where number of unique item is less than 3 with 1 and 0  
def replace_items(dataframe,column,variable_1,variable_2):
    """
    arguments :
    dataframe - the dataframe on which we want to perform the action 
    column - column of the dataframe where no of unique valuie are less than 3 
    variable1, variable2 - name of the element to replace 

    this function returns dataframe  
    """
    try:
        df = dataframe.replace({column:{variable_1:1,variable_2:0}})

    except :
        print("specified arguments are not correct please check")

    finally:
        return df 

In [16]:
# change the column International plan columns value wth {yes  :1,No:0}
telecom_df = replace_items(telecom_df,'International plan','Yes','No')#call the function 

In [17]:
# change the column Voice mail plan value wth {yes  :1,No:0}
telecom_df = replace_items(telecom_df,'Voice mail plan','Yes','No')#call the function 

In [18]:
# change the column Churn  value wth {True  :1,False:0}
telecom_df = replace_items(telecom_df,'Churn',True,False)#call the function 

#### **Creating special variable**

In [19]:
# make an another variable whic counts the overall duration of calls of a customer in a single 24 hour day 
telecom_df["overall_call_duration"] = telecom_df['Total day minutes']+telecom_df['Total eve minutes'] +telecom_df['Total night minutes']

In [20]:
# make an another variable whic counts the overall charges of calls of a customer in a single 24 hour day 
telecom_df["overall_call_charges"] = telecom_df['Total day charge']+telecom_df['Total eve charge'] +telecom_df['Total night charge']

In [21]:
# make an another variable whic counts the overall duration of a customer in a single 24 hour day 
telecom_df["overall_call_made"] = telecom_df['Total day calls']+telecom_df['Total eve calls'] +telecom_df['Total night calls']

In [22]:
#calculate the average plan f each customer 
telecom_df["Average_price_plan"] = telecom_df["overall_call_duration"]//telecom_df["overall_call_charges"]

In [46]:
# for the customers who become active at night 
telecom_df['intern_call_duration']=telecom_df['Total intl minutes']/telecom_df['Total intl calls']

In [ ]:
#Does the customer who have different night day and evening plan are more prone to churn 
telecom_df['international_rate_per_min']=df_churn_intl_yes['Total intl charge']/df_churn_intl_yes['Total intl minutes']
telecom_df['day_rate_per_min']=df_churn_intl_yes['Total day charge']/df_churn_intl_yes['Total day minutes']
telecom_df['eve_rate_per_min']=df_churn_intl_yes['Total eve charge']/df_churn_intl_yes['Total eve minutes']
telecom_df['night_rate_per_min']=df_churn_intl_yes['Total night charge']/df_churn_intl_yes['Total night minutes']

# **Dataset Wrangling**

In [24]:
#find out ratio of churn i.e peole who changed telecom to the people who are consistent 
print( {telecom_df["Churn"].value_counts().loc[0]/telecom_df["Churn"].value_counts().loc[1]})

{5.900621118012422}


**So for every 7 people 1 person has changed his operator**

####  **With Area code**

**Since all the datset has been divided in only three area code**

**We will begin our analysis with Area code**

In [25]:
telecom_df.groupby("Area code")["overall_call_duration"].mean().reset_index() # Average call duration in each area code 

Area code  overall_call_duration
0        408             577.674344
1        415             584.282296
2        510             580.340595

**Across all area code same amount of call_duration are made**

In [27]:
telecom_df.groupby("Area code")["Average_price_plan"].mean().reset_index() #across al the area code the average plan are same

Area code  Average_price_plan
0        408            9.887828
1        415            9.844713
2        510            9.878571

In [28]:
telecom_df.groupby("Area code")["Churn"].value_counts().unstack() # total sum of churn across all area code

Churn         0    1
Area code           
408         716  122
415        1419  236
510         715  125

**In 415 area code a double of customer moved from one telecom company to other than other two area code **

#### **With States**

In [34]:
print(telecom_df["State"].unique())# show unique states in the data 
print("")
print( {telecom_df["State"].nunique()}) # count of total unique states in data

['KS' 'OH' 'NJ' 'OK' 'AL' 'MA' 'MO' 'LA' 'WV' 'IN' 'RI' 'IA' 'MT' 'NY'
 'ID' 'VT' 'VA' 'TX' 'FL' 'CO' 'AZ' 'SC' 'NE' 'WY' 'HI' 'IL' 'NH' 'GA'
 'AK' 'MD' 'AR' 'WI' 'OR' 'MI' 'DE' 'UT' 'CA' 'MN' 'SD' 'NC' 'WA' 'NM'
 'NV' 'DC' 'KY' 'ME' 'MS' 'TN' 'PA' 'CT' 'ND']

{51}


In [41]:
telecom_df.groupby("State")["Churn"].sum().reset_index().sort_values("Churn",ascending = False) #arrange the index in their sum of largest to rarest churn customers 

State  Churn
31    NJ     18
43    TX     18
20    MD     17
22    MI     16
23    MN     15
34    NY     15
26    MT     14
33    NV     14
25    MS     14
40    SC     14
47    WA     14
16    KS     13
21    ME     13
6     CT     12
37    OR     11
27    NC     11
19    MA     11
2     AR     11
44    UT     10
49    WV     10
35    OH     10
36    OK      9
30    NH      9
50    WY      9
4     CA      9
15    IN      9
5     CO      9
13    ID      9
8     DE      9
9     FL      8
1     AL      8
17    KY      8
46    VT      8
41    SD      8
38    PA      8
10    GA      8
48    WI      7
24    MO      7
39    RI      6
32    NM      6
28    ND      6
42    TN      5
7     DC      5
14    IL      5
45    VA      5
29    NE      5
18    LA      4
3     AZ      4
11    HI      3
12    IA      3
0     AK      3

**`NJ` and `TX` has the largest churn customer . Also `AK` ,`iA` and `HI` has the lowest churn customer**

#### **With Calls and Their duration**

In [45]:
telecom_df.describe()

Account length    Area code  International plan  Voice mail plan  \
count     3333.000000  3333.000000         3333.000000      3333.000000   
mean       101.064806   437.182418            0.096910         0.276628   
std         39.822106    42.371290            0.295879         0.447398   
min          1.000000   408.000000            0.000000         0.000000   
25%         74.000000   408.000000            0.000000         0.000000   
50%        101.000000   415.000000            0.000000         0.000000   
75%        127.000000   510.000000            0.000000         1.000000   
max        243.000000   510.000000            1.000000         1.000000   

       Number vmail messages  Total day minutes  Total day calls  \
count            3333.000000        3333.000000      3333.000000   
mean                8.099010         179.775098       100.435644   
std                13.688365          54.467389        20.069084   
min                 0.000000           0.000000         0.000000   
25%                 0.000000         143.700000        87.000000   
50%                 0.000000         179.400000       101.000000   
75%                20.000000         216.400000       114.000000   
max                51.000000         350.800000       165.000000   

       Total day charge  Total eve minutes  Total eve calls  ...  \
count       3333.000000        3333.000000      3333.000000  ...   
mean          30.562307         200.980348       100.114311  ...   
std            9.259435          50.713844        19.922625  ...   
min            0.000000           0.000000         0.000000  ...   
25%           24.430000         166.600000        87.000000  ...   
50%           30.500000         201.400000       100.000000  ...   
75%           36.790000         235.300000       114.000000  ...   
max           59.640000         363.700000       170.000000  ...   

       Total intl minutes  Total intl calls  Total intl charge  \
count         3333.000000       3333.000000        3333.000000   
mean            10.237294          4.479448           2.764581   
std              2.791840          2.461214           0.753773   
min              0.000000          0.000000           0.000000   
25%              8.500000          3.000000           2.300000   
50%             10.300000          4.000000           2.780000   
75%             12.100000          6.000000           3.270000   
max             20.000000         20.000000           5.400000   

       Customer service calls        Churn  overall_call_duration  \
count             3333.000000  3333.000000            3333.000000   
mean                 1.562856     0.144914             581.627483   
std                  1.315491     0.352067              89.969330   
min                  0.000000     0.000000             275.200000   
25%                  1.000000     0.000000             521.000000   
50%                  1.000000     0.000000             583.100000   
75%                  2.000000     0.000000             642.500000   
max                  9.000000     1.000000             876.900000   

       overall_call_charges  overall_call_made  Average_price_plan  \
count           3333.000000        3333.000000         3333.000000   
mean              56.685173         300.657666            9.864086   
std               10.487816          34.328232            0.999915   
min               19.980000         188.000000            7.000000   
25%               49.590000         278.000000            9.000000   
50%               56.630000         301.000000           10.000000   
75%               63.650000         324.000000           10.000000   
max               92.560000         410.000000           15.000000   

       international_price_plan  
count                    3315.0  
mean                        0.0  
std                         0.0  
min                         0.0  
25%                         0.0  
50%                         0.0  
75%                

#**Data Visualization**